code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,194 @@
+	density = TRUE
+	anchored = TRUE
+	/// amount of biomass in the compost bin
+	var/biomass = 0
+	/// amount of compost in the compost bin
+	var/compost = 0
+	/// The maximum amount of biomass the compost bin can store.
+	var/biomass_capacity = BASE_BIOMASS_CAPACITY
+	/// The maximum amount of compost the compost bin can store.
+	var/compost_capacity = BASE_COMPOST_CAPACITY
+	var/composting = FALSE
+
+/obj/machinery/compost_bin/Initialize(mapload)
+	// If the bin is made with biomass in it try to compost it
+	if(biomass > 0)
+		compost()
+	return ..()
+
+
+/obj/machinery/compost_bin/on_deconstruction()
+	// returns wood instead of the non-existent components
+	new /obj/item/stack/sheet/wood(loc, 10)
+	return ..()
+
+/obj/machinery/compost_bin/screwdriver_act(mob/living/user, obj/item/I)
+	// there are no screws either
+	to_chat(user, "<span class='warning'>[src] has no screws!</span>")
+	return TRUE
+
+/obj/machinery/compost_bin/crowbar_act(mob/living/user, obj/item/I)
+	// no panel either
+	return default_deconstruction_crowbar(user, I, ignore_panel = TRUE)
+
+
+// accepts inserted plants and converts to biomass
+/obj/machinery/compost_bin/proc/make_biomass(obj/item/reagent_containers/food/snacks/grown/O)
+	// calculate biomass from plant nutriment and plant matter
+	var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
+	biomass += clamp(plant_biomass * 10, 1, biomass_capacity-biomass)
+	//plant delenda est
+	qdel(O)
+
+
+// takes care of plant insertion and conversion to biomass, and start composting what was inserted
+/obj/machinery/compost_bin/attackby(obj/item/O, mob/user, params)
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+
+	if(istype(O, /obj/item/storage/bag/plants))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more biomass!</span>")
+			return
+
+		var/obj/item/storage/bag/plants/PB = O
+		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
+
+			if(biomass >= biomass_capacity)
+				break
+			PB.remove_from_storage(G, src)
+			make_biomass(G)
+
+		// start composting after plants are inserted
+		compost()
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
+
+		SStgui.update_uis(src)
+		update_icon_state()
+		return TRUE
+
+	if(istype(O, /obj/item/reagent_containers/food/snacks/grown))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more plants!</span>")
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		make_biomass(O)
+		// start composting after plants are inserted
+		compost()
+		to_chat(user, "<span class='info'>You put [O] in [src].</span>")
+		SStgui.update_uis(src)
+		update_icon_state()
+		return TRUE
+
+	to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
```

Comment:
Added the return..(), and kept the message for now. Was there any reason for removing it, or was it a mistake?

Reply:
Its really just the standard to have this at the end, instead of the beginning, since you dont have to check for intent, and potentially mess up other items that arent considered in this scope

Reply:
>Its really just the standard to have this at the end, instead of the beginning

It's at the very end of the proc, I'm not sure I understand what you mean.

>since you dont have to check for intent, and potentially mess up other items that arent considered in this scope

I see. But I do want to check for intent, just for consistency's sake.

Reply:
Basically, what we want is this for attackby checks

 I guess the intent check is fine so that I guess you can break it with like deathnettles if you really wanted to
```
special cases (like plant bags, and food)
..() to return for all other cases
```

Reply:
So this should be fine?

>to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
return ..()

Reply:
Nah, keep the to_chat gone. If theyre attacking it with something else, its likely on purpose

Reply:
I just remembered why it was written as it was to begin with. I don't the machine to be attacked unless on harm intent, but when on harm intent It gets attacked by anything. 
Again this is for the sake of consistency, since the compost bin serves the same purpose of the biogenerator.

I tested again, just to be sure, and it does work as intended.

Sorry for all the mess there.